src/ipc/ipc-common.c

```diff
@@ -79,19 +79,17 @@ int ipc_process_on_core(uint32_t core, bool blocking)
 }
 
 /*
- * Components, buffers and pipelines all use the same set of monotonic ID
- * numbers passed in by the host. They are stored in different lists, hence
- * more than 1 list may need to be searched for the corresponding component.
+ * Components, buffers and pipelines are stored in the same lists, hence
+ * type and ID have to be used for the identification.
  */
-
-struct ipc_comp_dev *ipc_get_comp_by_id(struct ipc *ipc, uint32_t id)
+struct ipc_comp_dev *ipc_get_comp_dev(struct ipc *ipc, uint16_t type, uint32_t id)
 {
 	struct ipc_comp_dev *icd;
 	struct list_item *clist;
 
 	list_for_item(clist, &ipc->comp_list) {
 		icd = container_of(clist, struct ipc_comp_dev, list);
-		if (icd->id == id)
+		if ((icd->type & type) && icd->id == id)
```

Comment:
No, the value of icd->type could be one of 1,2,4, if type=COMPONENT (value is 1), you only get component, because 2 & 1 is false, 4 & 1 is also false. but if we want to get multiple type with `COMPONENT | BUFFER`, it is also allowed with above code, this is to mitigate IPC3, in pipeline connection, it want buffer or component. see below comments.

Reply:
I'm not sure what the correct logic should be, but the `if (icd->type & type)` looks to be too open?
If you have 3 components with:
1. `BUFFER`, id = 0
2. `BUFFER | COMPONENT`, id = 0
3. `BUFFER | PIPELINE`, id = 0
Is this valid?
you want to get a `BUFFER` for id = 0: all three matches, right?
 you want to get a `BUFFER | COMPONENT` for id = 0: all three matches, right?
 you want to get a `BUFFER | PIPELINE` for id = 0: all three matches, right?
 you want to get a `COMPONENT` for id = 0: only 2. is matching?

etc.

Reply:
@ujfalusi @lyakh I will try a new fix by modifying `ipc_comp_connect`, although the fix is trivial, but direct.